Services/DataBaseService/MigrationEngine.cs

```diff
@@ -0,0 +1,225 @@
+            Dictionary<string, ExecutedScript> scriptsToWriteDown)
+        {
+            var insertRow = new StringBuilder();
+            insertRow.Append("INSERT INTO [dbo].[ExecutedScripts] ");
+            insertRow.Append("(FileName, ExecutionTime, Code) ");
+            insertRow.Append("VALUES (@fileName, @executionTime, @code);");
+
+            using (var conn = new SqlConnection(connectionString))
+            {
+                conn.Open();
+                SqlTransaction writingTransaction = conn.BeginTransaction();
+                foreach (var script in scriptsToWriteDown)
+                {
+                    try
+                    {
+                        using (SqlCommand command = new SqlCommand(insertRow.ToString(), conn, writingTransaction))
+                        {
+                            command.Parameters.AddWithValue("@fileName", script.Key);
+                            command.Parameters.AddWithValue("@executionTime", script.Value.ExecutionTime);
+                            command.Parameters.AddWithValue("@code", script.Value.ExecutedCode);
+                            command.ExecuteNonQuery();
+                        }
+                    }
+                    catch (SqlException e)
+                    {
+                        // TODO replace with logs
+                        Console.WriteLine(e.Message + 
+                            $"\n\tWarning!\n{script.Key} script cannot be marked as executed.");
+                        try
+                        {
+                            writingTransaction.Rollback();
+                            // TODO replace with logs
+                            Console.WriteLine("\tWriting rollback is successful.");
+                        }
+                        catch
+                        {
+                            // TODO replace with logs
+                            Console.WriteLine("\tCouldn't rollback writing.");
+                        }
+                        throw;
+                    }
+                }
+                writingTransaction.Commit();
+                writingTransaction.Dispose();
+            }
+        }
+
+        private List<string> GetScriptsToExecute(string connectionString)
+        {
+            var scriptsLocation = configuration.GetSection("Locations")["MigrationScripts"];
+            var allScripts = Directory.GetFiles(scriptsLocation, "*.sql");
+            var scriptsToExecute = new List<string>();
+            var selectFileName = "SELECT (FileName) FROM [dbo].[ExecutedScripts] WHERE (FileName = @fileName);";
+
+            using (var conn = new SqlConnection(connectionString))
+            {
+                conn.Open();
+                foreach (var fileName in allScripts)
+                {
+                    bool alreadyExecuted;
+                    try
+                    {
+                        using (SqlCommand command = new SqlCommand(selectFileName, conn))
```

Comment:
you still select scripts from DB in multiple queries, do it in one query

Reply:
Now it's only one Select, and then looking up in the resulted List with Contains